src/nvim/os/env.c

```diff
@@ -87,3 +97,598 @@ void os_get_hostname(char *hostname, size_t len)
+/// the length of "name/".  Otherwise return "pend".
+static char_u *remove_tail(char_u *p, char_u *pend, char_u *name)
+{
+  int len = (int)STRLEN(name) + 1;
+  char_u      *newend = pend - len;
+
+  if (newend >= p
+      && fnamencmp(newend, name, len - 1) == 0
+      && (newend == p || after_pathsep(p, newend)))
+    return newend;
+  return pend;
+}
+
+/// Vim's version of getenv().
+/// Special handling of $HOME, $VIM and $VIMRUNTIME, allowing the user to
+/// override the vim runtime directory at runtime.  Also does ACP to 'enc'
+/// conversion for Win32.
+/// @param name Name of environment variable to expand
+/// @param mustfree Ouput parameter for the caller to determine if they are
+///           responsible for releasing memory. Must be initialized to false
+///           by the caller in order to work properly.
+char_u *vim_getenv(char_u *name, bool *mustfree)
+{
+  char_u      *p;
+  char_u      *pend;
+  int vimruntime;
+
+
+  p = (char_u *)os_getenv((char *)name);
+  if (p != NULL && *p == NUL)       /* empty is the same as not set */
+    p = NULL;
+
+  if (p != NULL) {
+    return p;
+  }
+
+  vimruntime = (STRCMP(name, "VIMRUNTIME") == 0);
+  if (!vimruntime && STRCMP(name, "VIM") != 0)
+    return NULL;
+
+  /*
+   * When expanding $VIMRUNTIME fails, try using $VIM/vim<version> or $VIM.
+   * Don't do this when default_vimruntime_dir is non-empty.
+   */
+  if (vimruntime
+#ifdef HAVE_PATHDEF
+      && *default_vimruntime_dir == NUL
+#endif
+      ) {
+    p = (char_u *)os_getenv("VIM");
+    if (p != NULL && *p == NUL)             /* empty is the same as not set */
+      p = NULL;
+    if (p != NULL) {
+      p = vim_version_dir(p);
+      if (p != NULL)
+        *mustfree = true;
+      else
+        p = (char_u *)os_getenv("VIM");
+    }
+  }
+
+  /*
+   * When expanding $VIM or $VIMRUNTIME fails, try using:
+   * - the directory name from 'helpfile' (unless it contains '$')
+   * - the executable name from argv[0]
+   */
+  if (p == NULL) {
+    if (p_hf != NULL && vim_strchr(p_hf, '$') == NULL)
+      p = p_hf;
+    if (p != NULL) {
+      /* remove the file name */
+      pend = path_tail(p);
+
+      /* remove "doc/" from 'helpfile', if present */
+      if (p == p_hf)
+        pend = remove_tail(p, pend, (char_u *)"doc");
+
+      /* for $VIM, remove "runtime/" or "vim54/", if present */
+      if (!vimruntime) {
+        pend = remove_tail(p, pend, (char_u *)RUNTIME_DIRNAME);
+        pend = remove_tail(p, pend, (char_u *)VIM_VERSION_NODOT);
+      }
+
+      /* remove trailing path separator */
+      /* With MacOS path (with  colons) the final colon is required */
+      /* to avoid confusion between absolute and relative path */
+      if (pend > p && after_pathsep(p, pend))
+        --pend;
+
+      // check that the result is a directory name
+      // pend should always be greater than p, so casting to an unsigned type
+      // should be safe here
```

Comment:
this type of comment is valuable as a review annotation, but not really useful in the source code. If it's really true, add an assert :)